File: hospital-ai/frontend/src/types/prescription.ts
/** Prescription Agent types — physician-review treatment recommendations. Never a final prescription. */
import type { OrchestratorDebugInfo } from '@/types/aiOrchestrator';

export type ApprovalStatus = 'Approved' | 'Requires Physician Review' | 'Rejected';
export type InteractionLevel = 'Minor' | 'Moderate' | 'Major' | 'Critical';
export type AllergyStatus = 'Safe' | 'Warning' | 'Contraindicated';

export const APPROVAL_STATUS_TONE: Record<string, 'green' | 'blue' | 'amber' | 'red' | 'gray'> = {
  Approved: 'green',
  'Requires Physician Review': 'amber',
  Rejected: 'red',
};

export const INTERACTION_LEVEL_TONE: Record<string, 'green' | 'blue' | 'amber' | 'red' | 'gray'> = {
  Minor: 'blue',
  Moderate: 'amber',
  Major: 'red',
  Critical: 'red',
};

export const ALLERGY_STATUS_TONE: Record<string, 'green' | 'blue' | 'amber' | 'red' | 'gray'> = {
  Safe: 'green',
  Warning: 'amber',
  Contraindicated: 'red',
};

export interface MedicationRecommendation {
  condition: string;
  medication_name: string;
  drug_class: string;
  purpose: string;
  evidence_source: string;
  clinical_guideline: string;
  confidence: number;
  alternative_drugs: string[];
  expected_outcome: string;
}

export interface DrugInteraction {
  drug_a: string;
  drug_b: string;
  interaction_level: InteractionLevel;
  explanation: string;
  recommendation: string;
}

export interface AllergyCheckItem {
  medication_name: string;
  status: AllergyStatus;
  reason: string;
  cross_reactivity: string[];
}

export interface DosageRecommendation {
  medication_name: string;
  starting_dose: string;
  maintenance_dose: string;
  maximum_dose: string;
  dose_adjustment: string[];
  adjustment_factors: string[];
}

export interface TreatmentPlan {
  medication_plan: string[];
  lifestyle_advice: string[];
  monitoring_plan: string[];
  recommended_lab_tests: string[];
  recommended_imaging: string[];
  recommended_specialists: string[];
  follow_up_interval: string;
  emergency_advice: string;
}

export interface PrescriptionValidation {
  duplicate_drugs: string[];
  contraindications_found: string[];
  max_dose_exceeded: string[];
  allergy_conflicts: string[];
  drug_warnings: string[];
  confidence_score: number;
  approval_status: ApprovalStatus;
  notes: string[];
}

export interface PrescriptionResult {
  id: string;
  patient_id: string;
  diagnosis_result_id?: string | null;
  research_result_id?: string | null;
  target_conditions_json: string[];
  medication_recommendations_json: MedicationRecommendation[];
  drug_interactions_json: DrugInteraction[];
  allergy_checks_json: AllergyCheckItem[];
  dosage_recommendations_json: DosageRecommendation[];
  treatment_plan_json: TreatmentPlan;
  validation_summary_json: PrescriptionValidation;
  summary?: string | null;
  engine: string;
  status: string;
  error_message?: string | null;
  processing_time_ms?: number | null;
  created_at: string;
  updated_at?: string | null;
}

export interface PrescriptionStartResult {
  patient_id: string;
  diagnosis_result_id?: string | null;
  research_result_id?: string | null;
  status: string;
  processing_time_ms: number;
  summary: string;
  engine: string;
  warnings: string[];
  target_conditions: string[];
  medication_recommendations: MedicationRecommendation[];
  drug_interactions: DrugInteraction[];
  allergy_checks: AllergyCheckItem[];
  dosage_recommendations: DosageRecommendation[];
  treatment_plan: TreatmentPlan;
  validation: PrescriptionValidation;
  prescription_result: PrescriptionResult;
  ai_debug: OrchestratorDebugInfo[];
}

export interface PrescriptionHistoryItem {
  id: string;
  created_at: string;
  summary?: string | null;
  approval_status?: ApprovalStatus | null;
  target_conditions: string[];
  status: string;
}

export interface PrescriptionStatus {
  patient_id: string;
  has_result: boolean;
  status: string;
  approval_status?: ApprovalStatus | null;
  confidence_score?: number | null;
  target_conditions: string[];
  summary?: string | null;
  processing_time_ms?: number | null;
  created_at?: string | null;
}

File: hospital-ai/backend/app/schemas/prescription.py
"""Pydantic schemas for the Prescription Agent."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.ai_orchestrator import OrchestratorDebugInfoOut


class PrescriptionStartRequest(BaseModel):
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None
    research_result_id: Optional[UUID] = None


class MedicationRecommendationOut(BaseModel):
    condition: str
    medication_name: str
    drug_class: str
    purpose: str = ""
    evidence_source: str = ""
    clinical_guideline: str = ""
    confidence: float = 0.0
    alternative_drugs: List[str] = Field(default_factory=list)
    expected_outcome: str = ""


class DrugInteractionOut(BaseModel):
    drug_a: str
    drug_b: str
    interaction_level: str = "Minor"
    explanation: str = ""
    recommendation: str = ""


class AllergyCheckItemOut(BaseModel):
    medication_name: str
    status: str = "Safe"
    reason: str = ""
    cross_reactivity: List[str] = Field(default_factory=list)


class DosageRecommendationOut(BaseModel):
    medication_name: str
    starting_dose: str = ""
    maintenance_dose: str = ""
    maximum_dose: str = ""
    dose_adjustment: List[str] = Field(default_factory=list)
    adjustment_factors: List[str] = Field(default_factory=list)


class TreatmentPlanOut(BaseModel):
    medication_plan: List[str] = Field(default_factory=list)
    lifestyle_advice: List[str] = Field(default_factory=list)
    monitoring_plan: List[str] = Field(default_factory=list)
    recommended_lab_tests: List[str] = Field(default_factory=list)
    recommended_imaging: List[str] = Field(default_factory=list)
    recommended_specialists: List[str] = Field(default_factory=list)
    follow_up_interval: str = ""
    emergency_advice: str = ""
    surgery_required: bool = False
    procedure_recommendations: List[str] = Field(default_factory=list)
    estimated_duration_minutes: Optional[int] = Field(default=None, ge=30, le=480)


class PrescriptionValidationOut(BaseModel):
    duplicate_drugs: List[str] = Field(default_factory=list)
    contraindications_found: List[str] = Field(default_factory=list)
    max_dose_exceeded: List[str] = Field(default_factory=list)
    allergy_conflicts: List[str] = Field(default_factory=list)
    drug_warnings: List[str] = Field(default_factory=list)
    confidence_score: float = 0.0
    approval_status: str = "Requires Physician Review"
    notes: List[str] = Field(default_factory=list)


class PrescriptionResultOut(BaseModel):
    id: UUID
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None
    research_result_id: Optional[UUID] = None
    target_conditions_json: List[str] = Field(default_factory=list)
    medication_recommendations_json: List[Any] = Field(default_factory=list)
    drug_interactions_json: List[Any] = Field(default_factory=list)
    allergy_checks_json: List[Any] = Field(default_factory=list)
    dosage_recommendations_json: List[Any] = Field(default_factory=list)
    treatment_plan_json: Dict[str, Any] = Field(default_factory=dict)
    validation_summary_json: Dict[str, Any] = Field(default_factory=dict)
    summary: Optional[str] = None
    engine: str = "rule_based"
    status: str = "Completed"
    error_message: Optional[str] = None
    processing_time_ms: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class PrescriptionStartResponse(BaseModel):
    patient_id: UUID
    diagnosis_result_id: Optional[UUID] = None
    research_result_id: Optional[UUID] = None
    status: str = "Completed"
    processing_time_ms: int
    summary: str
    engine: str
    warnings: List[str] = Field(default_factory=list)
    target_conditions: List[str] = Field(default_factory=list)
    medication_recommendations: List[MedicationRecommendationOut] = Field(default_factory=list)
    drug_interactions: List[DrugInteractionOut] = Field(default_factory=list)
    allergy_checks: List[AllergyCheckItemOut] = Field(default_factory=list)
    dosage_recommendations: List[DosageRecommendationOut] = Field(default_factory=list)
    treatment_plan: TreatmentPlanOut
    validation: PrescriptionValidationOut
    prescription_result: PrescriptionResultOut
    ai_debug: List[OrchestratorDebugInfoOut] = Field(default_factory=list)


class PrescriptionHistoryItemOut(BaseModel):
    id: UUID
    created_at: datetime
    summary: Optional[str] = None
    approval_status: Optional[str] = None
    target_conditions: List[str] = Field(default_factory=list)
    status: str = "Completed"


class PrescriptionStatusOut(BaseModel):
    """Lightweight status summary — mirrors the Intake Agent's status shape."""

    patient_id: UUID
    has_result: bool = False
    status: str = "Not Started"
    approval_status: Optional[str] = None
    confidence_score: Optional[float] = None
    target_conditions: List[str] = Field(default_factory=list)
    summary: Optional[str] = None
    processing_time_ms: Optional[int] = None
    created_at: Optional[datetime] = None
